### ISSUE TITLE
valid_ratio bug here

### ISSUE THREAD

**Issue by theodoruszq** [fundamentalvision/Deformable-DETR#36]
https://github.com/fundamentalvision/Deformable-DETR/blob/11169a60c33333af00a4849f1808023eba96a931/models/deformable_transformer.py#L249

The valid ratio has no effect. I consider it as a bug. Please check it and fix it if possible.

```
# use valid ratio
(Pdb) reference_points[0, :118, 0, 0]
tensor([0.0043, 0.0128, 0.0214, 0.0299, 0.0385, 0.0470, 0.0556, 0.0641, 0.0726,
        0.0812, 0.0897, 0.0983, 0.1068, 0.1154, 0.1239, 0.1325, 0.1410, 0.1496,
        0.1581, 0.1667, 0.1752, 0.1838, 0.1923, 0.2009, 0.2094, 0.2179, 0.2265,
        0.2350, 0.2436, 0.2521, 0.2607, 0.2692, 0.2778, 0.2863, 0.2949, 0.3034,
        0.3120, 0.3205, 0.3291, 0.3376, 0.3462, 0.3547, 0.3632, 0.3718, 0.3803,
        0.3889, 0.3974, 0.4060, 0.4145, 0.4231, 0.4316, 0.4402, 0.4487, 0.4573,
        0.4658, 0.4744, 0.4829, 0.4915, 0.5000, 0.5085, 0.5171, 0.5256, 0.5342,
        0.5427, 0.5513, 0.5598, 0.5684, 0.5769, 0.5855, 0.5940, 0.6026, 0.6111,
        0.6197, 0.6282, 0.6368, 0.6453, 0.6538, 0.6624, 0.6709, 0.6795, 0.6880,
        0.6966, 0.7051, 0.7137, 0.7222, 0.7308, 0.7393, 0.7479, 0.7564, 0.7650,
        0.7735, 0.7821, 0.7906, 0.7991, 0.8077, 0.8162, 0.8248, 0.8333, 0.8419,
        0.8504, 0.8590, 0.8675, 0.8761, 0.8846, 0.8932, 0.9017, 0.9103, 0.9188,
        0.9274, 0.9359, 0.9444, 0.9530, 0.9615, 0.9701, 0.9786, 0.9872, 0.9957,
        0.0043], device='cuda:0')
# not use valid_ratio
(Pdb) myreference_points[0, :118, 0, 0]
tensor([0.0043, 0.0128, 0.0214, 0.0299, 0.0385, 0.0470, 0.0556, 0.0641, 0.0726,
        0.0812, 0.0897, 0.0983, 0.1068, 0.1154, 0.1239, 0.1325, 0.1410, 0.1496,
        0.1581, 0.1667, 0.1752, 0.1838, 0.1923, 0.2009, 0.2094, 0.2179, 0.2265,
        0.2350, 0.2436, 0.2521, 0.2607, 0.2692, 0.2778, 0.2863, 0.2949, 0.3034,
        0.3120, 0.3205, 0.3291, 0.3376, 0.3462, 0.3547, 0.3632, 0.3718, 0.3803,
        0.3889, 0.3974, 0.4060, 0.4145, 0.4231, 0.4316, 0.4402, 0.4487, 0.4573,
        0.4658, 0.4744, 0.4829, 0.4915, 0.5000, 0.5085, 0.5171, 0.5256, 0.5342,
        0.5427, 0.5513, 0.5598, 0.5684, 0.5769, 0.5855, 0.5940, 0.6026, 0.6111,
        0.6197, 0.6282, 0.6368, 0.6453, 0.6538, 0.6624, 0.6709, 0.6795, 0.6880,
        0.6966, 0.7051, 0.7137, 0.7222, 0.7308, 0.7393, 0.7479, 0.7564, 0.7650,
        0.7735, 0.7821, 0.7906, 0.7991, 0.8077, 0.8162, 0.8248, 0.8333, 0.8419,
        0.8504, 0.8590, 0.8675, 0.8761, 0.8846, 0.8932, 0.9017, 0.9103, 0.9188,
        0.9274, 0.9359, 0.9444, 0.9530, 0.9615, 0.9701, 0.9786, 0.9872, 0.9957,
        0.0043], device='cuda:0')
```

```
# code
def get_reference_points(spatial_shapes, valid_ratios, device):
        """Core implementation of Deformable Transformer.
        """
        reference_points_list = []
        myreference_points_list = []
        for lvl, (H_, W_) in enumerate(spatial_shapes):
            ref_y, ref_x = torch.meshgrid(torch.linspace(0.5, H_ - 0.5, H_, dtype=torch.float32, device=device),
                                          torch.linspace(0.5, W_ - 0.5, W_, dtype=torch.float32, device=device))
           
            ref = torch.stack((_ref_x, _ref_y), -1)
            reference_points_list.append(ref)

            ref_y = ref_y.reshape(-1)[None] / H_
            ref_x = ref_x.reshape(-1)[None] / W_
            ref = torch.stack((ref_x, ref_y), -1)
            myreference_points_list.append(ref)

        reference_points = torch.cat(reference_points_list, 1)
        myreference_points = torch.cat(myreference_points_list, 1)

        reference_points = reference_points[:, :, None] * valid_ratios[:, None]
        myreference_points = myreference_points[:, :, None]
        diff = myreference_points - reference_points
        import pdb
        pdb.set_trace()
        return reference_points
```

**Comment by zehongs** [comment]
I think the reference_points are correct, which will be further processed and used in `grid_sample`. Computing reference_points doesn't need `valid_ratios`. I think it should be removed.

**Comment by weicheng113** [comment]
From the source code. It seems `valid_ratios` is not required.

Firstly, it is trying to scale up by `valid_radios`.
https://github.com/fundamentalvision/Deformable-DETR/blob/11169a60c33333af00a4849f1808023eba96a931/models/deformable_transformer.py#L244-L245

Next, it scales down by `valid_radios` before returning.
https://github.com/fundamentalvision/Deformable-DETR/blob/11169a60c33333af00a4849f1808023eba96a931/models/deformable_transformer.py#L249

Therefore, the two operation will cancel each other. I might write a test case to see if this is the case.

**Comment by cwei-bgl** [comment]
`
def get_reference_points(spatial_shapes, valid_ratios, device):
    reference_points_list = []
    for lvl, (H_, W_) in enumerate(spatial_shapes):

        ref_y, ref_x = torch.meshgrid(torch.linspace(0.5, H_ - 0.5, H_, dtype=torch.float32, device=device),
                                      torch.linspace(0.5, W_ - 0.5, W_, dtype=torch.float32, device=device))
        ref_y = ref_y.reshape(-1)[None] / (valid_ratios[:, None, lvl, 1] * H_)
        ref_x = ref_x.reshape(-1)[None] / (valid_ratios[:, None, lvl, 0] * W_)
        ref = torch.stack((ref_x, ref_y), -1)
        reference_points_list.append(ref)
    reference_points = torch.cat(reference_points_list, 1)
    reference_points = reference_points[:, :, None] * valid_ratios[:, None]
    return reference_points


def get_reference_points2(spatial_shapes, device):
    reference_points_list = []
    for lvl, (H_, W_) in enumerate(spatial_shapes):

        ref_y, ref_x = torch.meshgrid(torch.linspace(0.5, H_ - 0.5, H_, dtype=torch.float32, device=device),
                                      torch.linspace(0.5, W_ - 0.5, W_, dtype=torch.float32, device=device))
        ref_y = ref_y.reshape(-1)[None] / H_
        ref_x = ref_x.reshape(-1)[None] / W_
        ref = torch.stack((ref_x, ref_y), -1)
        reference_points_list.append(ref)
    reference_points = torch.cat(reference_points_list, 1)
    reference_points = reference_points[:, :, None]
    return reference_points


def test_get_reference_points():
    valid_ratios = torch.tensor(data=[0.9, 0.8], dtype=torch.float32).reshape(1, 1, 2)
    device = torch.device("cpu")
    spatial_shapes = torch.tensor(data=[5, 10]).reshape(1, 2)
    reference_points = get_reference_points(spatial_shapes=spatial_shapes, valid_ratios=valid_ratios, device=device)
    reference_points2 = get_reference_points2(spatial_shapes=spatial_shapes, device=device)
    assert torch.all(torch.isclose(reference_points, reference_points2))
`
From the test case above, the two methods produce the same result. So it looks `valid_ratios` is not required here.

**Comment by weicheng113** [comment]
```python
def get_reference_points(spatial_shapes, valid_ratios, device):
    reference_points_list = []
    for lvl, (H_, W_) in enumerate(spatial_shapes):

        ref_y, ref_x = torch.meshgrid(torch.linspace(0.5, H_ - 0.5, H_, dtype=torch.float32, device=device),
                                      torch.linspace(0.5, W_ - 0.5, W_, dtype=torch.float32, device=device))
        ref_y = ref_y.reshape(-1)[None] / (valid_ratios[:, None, lvl, 1] * H_)
        ref_x = ref_x.reshape(-1)[None] / (valid_ratios[:, None, lvl, 0] * W_)
        ref = torch.stack((ref_x, ref_y), -1)
        reference_points_list.append(ref)
    reference_points = torch.cat(reference_points_list, 1)
    reference_points = reference_points[:, :, None] * valid_ratios[:, None]
    return reference_points


def get_reference_points2(spatial_shapes, device):
    reference_points_list = []
    for lvl, (H_, W_) in enumerate(spatial_shapes):

        ref_y, ref_x = torch.meshgrid(torch.linspace(0.5, H_ - 0.5, H_, dtype=torch.float32, device=device),
                                      torch.linspace(0.5, W_ - 0.5, W_, dtype=torch.float32, device=device))
        ref_y = ref_y.reshape(-1)[None] / H_
        ref_x = ref_x.reshape(-1)[None] / W_
        ref = torch.stack((ref_x, ref_y), -1)
        reference_points_list.append(ref)
    reference_points = torch.cat(reference_points_list, 1)
    reference_points = reference_points[:, :, None]
    return reference_points


def test_get_reference_points():
    valid_ratios = torch.tensor(data=[0.9, 0.8], dtype=torch.float32).reshape(1, 1, 2)
    device = torch.device("cpu")
    spatial_shapes = torch.tensor(data=[5, 10]).reshape(1, 2)
    reference_points = get_reference_points(spatial_shapes=spatial_shapes, valid_ratios=valid_ratios, device=device)
    reference_points2 = get_reference_points2(spatial_shapes=spatial_shapes, device=device)
    assert torch.all(torch.isclose(reference_points, reference_points2))  # assert passed. The two arrays are equal.
```

From the test case above, the two methods produce the same result. So it looks valid_ratios is not required here.

**Comment by Vincent-luo** [comment]
@weicheng113 you didn't consider the layer demension, actually the demension of `reference_points` returned by `get_reference_points` and `get_reference_points2` is not the same, the reason is the elementwise multiplication `reference_points[:, :, None] * valid_ratios[:, None]` will do broadcasting internally.

You can refer to [this](https://github.com/open-mmlab/mmdetection/issues/8656#issuecomment-1295866151), I think it explains very well. 

TL;DR
The author's implementation has nothing wrong and it's not a bug.
